File: aai/identifyArtifacts/services/openai_service.py
import os
import requests
from dotenv import load_dotenv
from identifyArtifacts.services.expert_service import AIService
from identifyArtifacts.utilities.image_encoder import encodeImage
from identifyArtifacts.prompts import OPENAIAPI_PROMPT_TEMPLATE

class OpenAIService(AIService):
    def __init__(self, api_key: str = None):
        load_dotenv()  # Load environment variables from .env file
        self.api_key = api_key or os.getenv("OPENAI_API_KEY")
        if not self.api_key:
            raise ValueError("OpenRouter API Key not found. Please set OPENROUTER_API_KEY in your environment.")
        
    def get_artifact_description(self) -> str:
        load_dotenv()
        #prompt = "The artifact is sharp. It is small, almost the size of a hand, and is black in colour. It also looks like an arrow."
        current_dir = os.path.dirname(__file__)
        img_path = os.path.join(current_dir, "s-l1200.jpg")
        #img_path = "s-l1200.jpg"  

        # Getting the Base64 string
        image = encodeImage(img_path)
        base64_image = encodeImage(img_path)

        # OpenAI API Key
        api_key = os.environ['OPENAI_API_KEY']

        prompt = OPENAIAPI_PROMPT_TEMPLATE


        headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}"
        }

        payload = {
        "model": "gpt-4o-mini",
        "messages": [
            {
            "role": "user",
            "content": [
                {
                "type": "text",
                "text": prompt
                },
                {
                "type": "image_url",
                "image_url": {
                    "url": f"data:image/jpeg;base64,{base64_image}"
                }
                }
            ]
            }
        ],
        "max_tokens": 3000
        }

        response = requests.post("https://api.openai.com/v1/chat/completions", headers=headers, json=payload)

        # Convert the response to JSON
        json_data = response.json()

        # Extract the text from 'content'
        try:
            content = json_data["choices"][0]["message"]["content"]
        except (KeyError, IndexError):
            content = "Error: Unexpected response structure"


        print(content)

File: androidapp/app/services/openai_service.js
import axios from 'axios';
import { OPENAIAPI_PROMPT_TEMPLATE } from '../utilities/prompts.js'; // Assuming this is the correct path

class OpenAIService {
    async getArtifactDescription(apiKey, imageURL) {
        const prompt = OPENAIAPI_PROMPT_TEMPLATE;
        
        const headers = {
            "Content-Type": "application/json",
            "Authorization": `Bearer ${apiKey}`
        };

        const payload = {
            model: "gpt-4o-mini",  // Ensure that the model supports image processing (if applicable)
            messages: [
                {
                    role: "user",
                    content: prompt  // Text-based prompt for artifact description
                },
                {
                    role: "user",
                    content: imageURL  // Image URL
                }
            ],
            max_tokens: 3000
        };

        try {
            const response = await axios.post("https://api.openai.com/v1/chat/completions", payload, { headers });

            // Extracting the content from the API response
            const content = response.data.choices[0]?.message?.content || "Error: Unexpected response structure";

            return content;
        } catch (error) {
            console.error("Error during API call:", error.response?.data || error.message);
            return "Error: Unable to fetch the response from OpenAI.";
        }
    }
}

export { OpenAIService };
